pkg/persistence/disk_persistence.go

```diff
@@ -106,15 +103,18 @@ func (ds *diskPersistence) Snapshot(data []byte, dirName, fileName string) error
 	}
 
 	dirPath := fmt.Sprintf("%s/%s", ds.dataDir, snapshotDir)
-	err := createDir(dirPath, dirName)
+	err := ensureDirectoryExists(dirPath, dirName)
 	if err != nil {
 		return err
 	}
 
 	filePath := fmt.Sprintf("%s/%s/%s", dirPath, dirName, fileName+snapshotSuffix)
 
+	ds.snapshotMutex.Lock()
```

Comment:
Shouldn't `ensureDirectoryExist` be included in the critical section guarded by this mutex? There is a possibility we can have two concurrent goroutines continuing because the directory does not exist and one of them will fail later on `os.Mkdir`.

Reply:
Good point! Seems it should.